File: PitBoard(Android)/app/src/main/java/com/pitboard/app/util/DateTimeFormatters.kt
package com.pitboard.app.util

import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Locale

object DateTimeFormatters {

    private val syncFormatter = DateTimeFormatter.ofPattern("d MMM yyyy, HH:mm", Locale("es", "ES"))
    private val eventDateFormatter = DateTimeFormatter.ofPattern("EEE d MMM · HH:mm", Locale("es", "ES"))
    private val timeOnlyFormatter = DateTimeFormatter.ofPattern("HH:mm", Locale("es", "ES"))
    private val eventDateFormatterLong = DateTimeFormatter.ofPattern("EEEE d 'de' MMMM 'de' yyyy, HH:mm", Locale("es", "ES"))

    fun formatSyncTimestamp(lastSyncedAt: Long?, importedAt: Long): String {
        val isImportOnly = lastSyncedAt == null
        val millis = lastSyncedAt ?: importedAt
        val zone = ZoneId.systemDefault()
        val zoned = Instant.ofEpochMilli(millis).atZone(zone)
        val now = Instant.now().atZone(zone)
        val prefix = if (isImportOnly) "Importado" else "Actualizado"
        val absolute = syncFormatter.format(zoned)

        val minutesAgo = ChronoUnit.MINUTES.between(zoned, now)
        if (isImportOnly) return "$prefix $absolute"

        return when {
            minutesAgo < 1 -> "$prefix hace un momento"
            minutesAgo < 60 -> "$prefix hace $minutesAgo min"
            ChronoUnit.HOURS.between(zoned, now) < 24 -> {
                val hours = ChronoUnit.HOURS.between(zoned, now)
                "$prefix hace $hours h"
            }
            ChronoUnit.DAYS.between(zoned.toLocalDate(), now.toLocalDate()) == 1L ->
                "$prefix ayer, $absolute"
            else -> "$prefix $absolute"
        }
    }

    fun formatEventDateTime(startTimeUtc: Long): String {
        val zoned = Instant.ofEpochMilli(startTimeUtc).atZone(ZoneId.systemDefault())
        return eventDateFormatter.format(zoned)
    }

    /** Solo la hora, en la zona horaria local del dispositivo (ej: "15:00"). Usado para
     *  mostrar la hora exacta de inicio en las notificaciones. */
    fun formatTimeOnly(startTimeUtc: Long): String {
        val zoned = Instant.ofEpochMilli(startTimeUtc).atZone(ZoneId.systemDefault())
        return timeOnlyFormatter.format(zoned)
    }

    /** Fecha y hora completas (día de la semana, día, mes y año) en la zona horaria del
     *  DISPOSITIVO — para el panel de detalle de un evento (ver EventsScreen), donde
     *  "EEE d MMM · HH:mm" (formatEventDateTime) se queda corto porque no lleva el año. */
    fun formatEventDateTimeLong(startTimeUtc: Long): String {
        val zoned = Instant.ofEpochMilli(startTimeUtc).atZone(ZoneId.systemDefault())
        return eventDateFormatterLong.format(zoned).replaceFirstChar { it.uppercase() }
    }

    /** Igual que [formatEventDateTimeLong] pero en la zona horaria indicada (ej. la del
     *  circuito, "America/New_York") en vez de la del dispositivo — null si el id de zona no
     *  es válido, para no mostrar una hora inventada. */
    fun formatEventDateTimeInZone(startTimeUtc: Long, zoneId: String): String? {
        val zone = runCatching { ZoneId.of(zoneId) }.getOrNull() ?: return null
        val zoned = Instant.ofEpochMilli(startTimeUtc).atZone(zone)
        return eventDateFormatterLong.format(zoned).replaceFirstChar { it.uppercase() }
    }
}

File: PitBoard(Iphone)/PitBoardKit/Sources/PitBoardKit/Support/DateTimeFormatters.swift
import Foundation

/// Equivalente de `DateTimeFormatters.kt`. `formatSyncTimestamp` no se porta — en Android
/// ya no la usa nadie (era del antiguo calendario `.ics` importado a mano, eliminado antes
/// de esta migración); `CategoryStandingsScreen.kt` tiene su propio `formatLastUpdated`
/// local con `SimpleDateFormat`, que aquí sí se porta como `formatLastUpdated`.
public enum DateTimeFormatters {

    private static func formatter(_ pattern: String, timeZone: TimeZone? = nil) -> DateFormatter {
        let f = DateFormatter()
        f.locale = Locale(identifier: "es_ES")
        f.dateFormat = pattern
        if let timeZone { f.timeZone = timeZone }
        return f
    }

    private static let eventDateFormatter = formatter("EEE d MMM · HH:mm")
    private static let timeOnlyFormatter = formatter("HH:mm")
    private static let eventDateFormatterLong = formatter("EEEE d 'de' MMMM 'de' yyyy, HH:mm")
    private static let lastUpdatedFormatter = formatter("d MMM, HH:mm")

    /// Usa la zona horaria del DISPOSITIVO (comportamiento por defecto de `DateFormatter`).
    public static func formatEventDateTime(_ date: Date) -> String {
        eventDateFormatter.string(from: date)
    }

    /// Solo la hora, en la zona horaria local del dispositivo — usado en las notificaciones.
    public static func formatTimeOnly(_ date: Date) -> String {
        timeOnlyFormatter.string(from: date)
    }

    /// Fecha y hora completas (día de la semana, día, mes y año) en la zona del dispositivo.
    public static func formatEventDateTimeLong(_ date: Date) -> String {
        capitalizeFirst(eventDateFormatterLong.string(from: date))
    }

    /// Igual que `formatEventDateTimeLong` pero en la zona indicada (ej. la del circuito,
    /// "America/New_York") — nil si el id de zona no es válido, para no mostrar una hora
    /// inventada.
    public static func formatEventDateTime(_ date: Date, inZone zoneId: String) -> String? {
        guard let timeZone = TimeZone(identifier: zoneId) else { return nil }
        let f = formatter("EEEE d 'de' MMMM 'de' yyyy, HH:mm", timeZone: timeZone)
        return capitalizeFirst(f.string(from: date))
    }

    public static func formatLastUpdated(_ date: Date) -> String {
        lastUpdatedFormatter.string(from: date)
    }

    private static func capitalizeFirst(_ s: String) -> String {
        guard let first = s.first else { return s }
        return first.uppercased() + s.dropFirst()
    }
}

/// Los puntos vienen como `Double` (necesario para NASCAR/IndyCar), pero se muestran sin
/// decimales cuando son un número entero — que es lo habitual. Equivalente exacto de
/// `formatPoints()` en `StandingsScreen.kt`.
public func formatPoints(_ points: Double) -> String {
    if points.truncatingRemainder(dividingBy: 1) == 0 && abs(points) < 1e15 {
        return String(Int64(points))
    }
    return String(points)
}
